### PR TITLE
Fix offset access + key-of template type combination

### PR DESCRIPTION
**Pull request by rvanvelzen**
Fix phpstan/phpstan#7788

### CODE REVIEW

File: src/Reflection/ResolvedMethodReflection.php [hunk 1]
```diff
@@ -39,9 +40,14 @@ public function getVariants(): array
 
 		$variants = [];
 		foreach ($this->reflection->getVariants() as $variant) {
+			$resolvedTemplateTypeMap = new TemplateTypeMap(array_merge(
+				$variant->getTemplateTypeMap()->getTypes(),
+				$this->resolvedTemplateTypeMap->getTypes(),
+			));
```
Comment (ondrejmirtes) [inline src/Reflection/ResolvedMethodReflection.php]: Can you please elaborate a bit what happens exactly and why is this needed? And BTW could the `$this->resolvedTemplateTypeMap` be already correct when this object is being constructed? :)

Comment (rvanvelzen) [reply src/Reflection/ResolvedMethodReflection.php]: So what happens is basically that `K` as a parameter disappears somewhere. I've spent a day tracking through all calls and am not sure what is happening, but this ensures that `K` is still present with the correct type in the template map.

This probably isn't the correct solution though, which is why I'm keeping it as a draft until I'm convinced that there is no other solution

Comment (ondrejmirtes) [reply src/Reflection/ResolvedMethodReflection.php]: Alright, so `@return T` works: https://phpstan.org/r/0891ac99-4e84-438e-86ef-b6f9f0df91c2

`@return K` is the problem, that's where `'value'` gets lost: https://phpstan.org/r/2d027c68-ed9b-4e0a-8126-11807e2d0a6e

Maybe it's note-worthy this doesn't work at all `K of 'test'|'value'` https://phpstan.org/r/acc2153f-90b4-4331-8979-7262073c63d2 because of `TemplateUnionType::shouldGeneralizeInferredType()` (it's the wrong abstraction for this, but probably not worth changing now if we want to get rid of generalization altogether).

Comment (rvanvelzen) [reply src/Reflection/ResolvedMethodReflection.php]: It took some time, but I realized the root cause was that resolving the template key-of type did not result in a new template type. PR updated with a working fix :tada:

**Comment by ondrejmirtes** [comment]
Perfect, thank you!
